grakn-engine/src/main/java/ai/grakn/engine/cache/EngineCacheDistributed.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.engine.cache;
+
+import ai.grakn.concept.ConceptId;
+import ai.grakn.engine.tasks.manager.ZookeeperConnection;
+import ai.grakn.exception.EngineStorageException;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static ai.grakn.engine.tasks.config.ZookeeperPaths.ENGINE_CACHE_JOBS;
+import static ai.grakn.engine.tasks.config.ZookeeperPaths.ENGINE_CACHE_KEYSPACES;
+import static java.lang.String.format;
+import static org.apache.commons.lang.SerializationUtils.deserialize;
+import static org.apache.spark.util.Utils.serialize;
+
+/**
+ * <p>
+ *     Engine's internal Concept ID cache
+ * </p>
+ *
+ * <p>
+ *    This is a cache which houses {@link org.apache.tinkerpop.gremlin.structure.Vertex} ids. It keeps these ids
+ *    so that we can lookup the vertices in need of postprocessing directly.
+ *
+ *    We cannot relay on {@link ai.grakn.concept.ConceptId}s because the indexed lookup maybe faulty with
+ *    vertices in need of post processing.
+ *
+ *    This distributed version persists everything using {@link ai.grakn.engine.tasks.manager.ZookeeperConnection}
+ * </p>
+ *
+ * @author fppt
+ */
+//TODO: Maybe we can merge this with Stand alone engine cache using Kafka for example?
+//TODO: This class may be in need of sever optimisation. Constantly serialising and deserialising maps and sets is likely to be slow.
+public class EngineCacheDistributed extends EngineCacheAbstract{
+    private static EngineCacheDistributed instance = null;
+    private final ZookeeperConnection zookeeper;
+
+
+    private EngineCacheDistributed(ZookeeperConnection zookeeper){
+        this.zookeeper = zookeeper;
+    }
+
+    public static synchronized EngineCacheDistributed init(ZookeeperConnection connection){
+        if(instance == null) instance = new EngineCacheDistributed(connection);
+        return instance;
```

Comment:
I will try to change that but I recall getting some epic failures on that change